src/Voron/Impl/EncryptionBuffersPool.cs

```diff
@@ -1,125 +1,290 @@
+            var numberOfSlots = Bits.MostSignificantBit(_maxBufferSizeToKeepInBytes) + 1;
+            _items = new ConcurrentStack<NativeAllocation>[numberOfSlots];
+
             for (int i = 0; i < _items.Length; i++)
             {
                 _items[i] = new ConcurrentStack<NativeAllocation>();
             }
 
             LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
 
-            _disposeOnceRunner = new DisposeOnce<SingleAttempt>(() =>
-            {
-                ReleaseUnmanagedResources();
-                GC.SuppressFinalize(this);
-            });
+            _cleanupTimer = new Timer(CleanupTimer, null, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1));
         }
 
         public byte* Get(int size, out NativeMemory.ThreadStats thread)
         {
             size = Bits.PowerOf2(size);
 
-            if (size > Constants.Size.Megabyte * 16)
+            if (size > _maxBufferSizeToKeepInBytes)
             {
                 // We don't want to pool large buffers
                 return PlatformSpecific.NativeMemory.Allocate4KbAlignedMemory(size, out thread);
             }
 
             var index = Bits.MostSignificantBit(size);
-
-            if (_items[index].TryPop(out var allocation))
+            while (_items[index].TryPop(out var allocation))
             {
-                thread = allocation.AllocatingThread;
-                return (byte*)allocation.Ptr;
+                if (allocation.InUse.Raise() == false)
+                    continue;
+
+                thread = NativeMemory.ThreadAllocations.Value;
+                thread.Allocations += size;
+                return allocation.Ptr;
             }
 
             return PlatformSpecific.NativeMemory.Allocate4KbAlignedMemory(size, out thread);
         }
 
-        public void Return(byte* ptr, long size, NativeMemory.ThreadStats allocatingThread)
+        public void Return(byte* ptr, long size, NativeMemory.ThreadStats allocatingThread, long generation)
         {
             if (ptr == null)
                 return;
 
             size = Bits.PowerOf2(size);
             Sodium.sodium_memzero(ptr, (UIntPtr)size);
 
-            if (size > Constants.Size.Megabyte * 16 || LowMemoryNotification.Instance.LowMemoryState)
+            if (size > _maxBufferSizeToKeepInBytes || 
+                _lowMemoryFlag.IsRaised() && generation < Generation)
             {
-                // We don't want to pool large buffers / clear them up on low memory
+                // - don't want to pool large buffers
+                // - release all the buffers that were created before we got the low memory event
                 PlatformSpecific.NativeMemory.Free4KbAlignedMemory(ptr, size, allocatingThread);
                 return;
             }
 
+            // updating the thread allocations since we released the memory back to the pool
+            NativeMemory.UpdateMemoryStatsForThread(allocatingThread, size);
```

Comment:
This is an issue about accounting.
From outside, we are always "charged" for this memory.
You are changing things so it will only account for the memory when it is out of the pool. But what about memory that is *in* the pool? That would be hidden from us.

Reply:
I'm updating it only for that thread and the total amount of allocated memory remains (only updated when we actually release the memory).

Reply:
The problem here is that you are not accounting allocated memory (from the OS perspective), which will make it harder for us to know that there is pooled memory that isn't in use

Reply:
The total unmanaged and the total size of the encryption buffer pool will be visible in the stats